File: omni/umbrella/tests/cpp/TestSimulatorInteraction.cpp
// SPDX-FileCopyrightText: Copyright (c) 2025-2026 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
// SPDX-License-Identifier: BSD-3-Clause
//
// Tests for interaction functions (user interactions, reset on stop, etc.)

#include <omni/physics/simulation/IPhysics.h>
#include "SimulationRegistry.h"
#include <doctest/doctest.h>

using namespace omni::physics;

//=============================================================================
// TEST: Interaction Functions
//=============================================================================
TEST_CASE("Interaction Functions")
{
    SimulationRegistry registry;

    SUBCASE("Reset on stop")
    {
        Simulation simulation;

        bool resetOnStopDisabled = false;
        simulation.interactionFns.disableResetOnStop = [&resetOnStopDisabled](bool disable) {
            resetOnStopDisabled = disable;
        };

        simulation.interactionFns.isDisabledResetOnStop = [&resetOnStopDisabled]() -> bool {
            return resetOnStopDisabled;
        };

        SimulationId simId = registry.registerSimulation(simulation, "InteractionTest");
        const Simulation* sim = registry.getSimulation(simId);

        // Test disable
        sim->interactionFns.disableResetOnStop(true);
        REQUIRE(sim->interactionFns.isDisabledResetOnStop() == true);

        // Test enable
        sim->interactionFns.disableResetOnStop(false);
        REQUIRE(sim->interactionFns.isDisabledResetOnStop() == false);

        registry.unregisterSimulation(simId);
    }

    SUBCASE("Handle raycast")
    {
        Simulation simulation;

        bool raycastHandled = false;
        simulation.interactionFns.handleRaycast = [&raycastHandled](
            const float* /*origin*/, const float* /*dir*/, bool /*input*/) {
            raycastHandled = true;
        };

        SimulationId simId = registry.registerSimulation(simulation, "RaycastInteractionTest");
        const Simulation* sim = registry.getSimulation(simId);

        float origin[3] = { 0.0f, 0.0f, 0.0f };
        float dir[3] = { 1.0f, 0.0f, 0.0f };
        sim->interactionFns.handleRaycast(origin, dir, true);

        REQUIRE(raycastHandled == true);

        registry.unregisterSimulation(simId);
    }
}

File: omni/umbrella/tests/python/testSimulatorInteraction.py
# SPDX-FileCopyrightText: Copyright (c) 2025-2026 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
# SPDX-License-Identifier: LicenseRef-NvidiaProprietary
#

import _carb_setup  # noqa: F401 - initializes Carbonite framework

import carb
from typing import Any
import unittest
from _physics import (
    Simulation,
    acquire_physics_interaction_interface,
    acquire_physics_interface,
    DebugDataItemType,
    k_invalid_simulation_id,
)


debug_data_entries = {
    "test float": {"type": DebugDataItemType.FLOAT, "value": 0.0},
    "test int": {"type": DebugDataItemType.INT, "value": 0},
    "test string": {"type": DebugDataItemType.STRING, "value": "test"},
    "test bool": {"type": DebugDataItemType.BOOL, "value": True},
    "test vector3": {"type": DebugDataItemType.VECTOR, "value": (1.0, 2.0, 3.0)},
    "test point3": {"type": DebugDataItemType.POINT, "value": (1.0, 2.0, 3.0)},
    "test quaternion": {"type": DebugDataItemType.QUATERNION, "value": (1.0, 2.0, 3.0, 4.0)},
}


class MockInteraction:
    def __init__(self):
        self.is_reset_on_stop_disabled = False
        self.raycast_count = 0
        self.last_raycast_input = False

    def disable_reset_on_stop(self, disable):
        self.is_reset_on_stop_disabled = disable

    def is_disabled_reset_on_stop(self):
        return self.is_reset_on_stop_disabled

    def handle_raycast(self, origin, direction, has_input):
        self.raycast_count += 1
        self.last_raycast_input = has_input
        return True

    def get_prim_debug_data(self, prim_path: str) -> dict[str, dict[str, Any]]:
        return debug_data_entries


class TestSimulateInteraction(unittest.TestCase):
    def setUp(self):
        self.physics = acquire_physics_interface()
        self.physics_interaction = acquire_physics_interaction_interface()
        self.mock = MockInteraction()

        self.simulation = Simulation()
        self.simulation.interaction_fns.disable_reset_on_stop = self.mock.disable_reset_on_stop
        self.simulation.interaction_fns.is_disabled_reset_on_stop = self.mock.is_disabled_reset_on_stop
        self.simulation.interaction_fns.handle_raycast = self.mock.handle_raycast
        self.simulation.interaction_fns.get_prim_debug_data = self.mock.get_prim_debug_data

        self.simulation_id = self.physics.register_simulation(self.simulation, "MockInteractionSim")
        self.assertNotEqual(self.simulation_id, k_invalid_simulation_id)

    def tearDown(self):
        if self.simulation_id != k_invalid_simulation_id:
            self.physics.unregister_simulation(self.simulation_id)

    def test_reset_on_stop_control(self):
        self.assertFalse(self.physics_interaction.is_disabled_reset_on_stop(self.simulation_id))

        self.physics_interaction.disable_reset_on_stop(True)
        self.assertTrue(self.mock.is_reset_on_stop_disabled)
        self.assertTrue(self.physics_interaction.is_disabled_reset_on_stop(self.simulation_id))

        self.physics_interaction.disable_reset_on_stop(False)
        self.assertFalse(self.mock.is_reset_on_stop_disabled)

    def test_raycast_handling(self):
        origin = carb.Float3(1.0, 2.0, 3.0)
        direction = carb.Float3(0.0, 1.0, 0.0)

        self.physics_interaction.handle_raycast(origin, direction, True)
        self.assertEqual(self.mock.raycast_count, 1)
        self.assertTrue(self.mock.last_raycast_input)

        self.physics_interaction.handle_raycast(origin, direction, False)
        self.assertEqual(self.mock.raycast_count, 2)
        self.assertFalse(self.mock.last_raycast_input)

    def test_get_prim_debug_data(self):
        debug_data = self.physics_interaction.get_prim_debug_data("/World/Cube")
        self.assertEqual(set(debug_data.keys()), set(debug_data_entries.keys()))
        for key, entry in debug_data.items():
            golden = debug_data_entries[key]
            for item_key, item_value in entry.items():
                self.assertEqual(item_value, golden[item_key],
                                 f"Mismatch for key {key}, item {item_key}")


if __name__ == "__main__":
    unittest.main()
